File: services/problems/src/worker/runner/javascript/runner.js
const fs = require('fs');

let userCode;
let functionName;

try {
  userCode = fs.readFileSync('./user_code.js', 'utf-8');
  functionName = fs.readFileSync('./function_name.txt', 'utf-8').trim();
} catch (e) {
  const error = {
    type: "FILE_READ_ERROR",
    message: e.message,
    stack: e.stack
  };
  console.error(JSON.stringify(error));
  process.exit(1);
}

try {
  eval(userCode);
} catch (e) {
  const error = {
    type: "SYNTAX_ERROR",
    message: e.message,
    stack: e.stack
  };
  console.error(JSON.stringify(error));
  process.exit(1);
}

try {
  const input = JSON.parse(process.argv[2]);
  const args = Array.isArray(input) ? input : [input];

  if (typeof global[functionName] !== 'function') {
    throw new Error(`Function "${functionName}" is not defined properly.`);
  }

  const result = global[functionName](...args);

  // Handle Promise-returning functions
  if (result instanceof Promise) {
    result
      .then(res => {
        console.log(JSON.stringify(res));
      })
      .catch(err => {
        const error = {
          type: "ASYNC_RUNTIME_ERROR",
          message: err.message,
          stack: err.stack
        };
        console.error(JSON.stringify(error));
        process.exit(1);
      });
  } else {
    console.log(JSON.stringify(result));
  }
} catch (e) {
  const error = {
    type: "RUNTIME_ERROR",
    message: e.message,
    stack: e.stack
  };
  console.error(JSON.stringify(error));
  process.exit(1);
}

File: services/problems/src/worker/runner/python/runner.py
import json
import sys

try:
    # Read user code
    with open('user_code.py', 'r', encoding='utf-8') as f:
        user_code = f.read()

    # Read function name
    with open('function_name.txt', 'r', encoding='utf-8') as f:
        function_name = f.read().strip()

    # Prepare an isolated namespace
    user_namespace = {}

    # Execute the user's code
    exec(user_code, user_namespace)

    # Read the input
    input_args = json.loads(sys.argv[1])

    # Ensure function exists
    if function_name not in user_namespace or not callable(user_namespace[function_name]):
        raise Exception(f"Function '{function_name}' not found in user_code.py")

    # Call the function
    result = user_namespace[function_name](*input_args)

    # Return result
    print(json.dumps({ "result": result }))

except Exception as e:
    # Return error message in JSON format
    print(json.dumps({ "error": str(e) }))
    sys.exit(1)
